### PR TITLE
Cache improvements

### PR DESCRIPTION
**Pull request by jschr**
Caching improvements that make a lot more content cache-able by default: 

1. Cache responses even if the Cache-Control header is set to max-age=0. Note that the max-age=0 header is still respected while online but it's now allowed to use the cache if the device is offline. 
2. Cache 403, 404 and 429 errors as these may be intermittent errors that seem ok to cache by default. In the future we could consider allowing Raydiant apps to customize the cache behaviour.

### CODE REVIEW

File: Dockerfile.build [hunk 1]
```diff
@@ -17,7 +17,7 @@ COPY certs ./certs
 COPY conf ./conf
 COPY ngx_http_proxy_connect_module ./ngx_http_proxy_connect_module
 COPY scripts ./scripts
-COPY allow_options_cache.patch .
+COPY patches ./patches
```
Comment (jschr) [inline Dockerfile.build:20]: Added all our ngnix patches to patches dir

File: test/test.py [hunk 2]
```diff
@@ -150,7 +151,7 @@ def test_get_rss(self):
         self.case_fetch_resource(RSS_URL, eta=5)
 
     def test_get_hls(self):
-        self.case_fetch_uncached_resource(HLS_STREAM_URL)
+        self.case_fetch_resource(HLS_STREAM_URL)
```
Comment (jschr) [inline test/test.py:154]: The HLS test feed is now cache-able by default with these changes.

**Review by brietaylor — APPROVED** [review]
Looks good to me.  Since we're now caching no-store, private, max-age=0, and errors; are we just trying to cache everything?  Are there any things that we want to exclude?

Also, I'm wondering what the retry interval looks like for 5xx and 429 errors.  If there's a single intermittent error how long does it take to recover and retry?

**Comment by jschr** [comment]
> Since we're now caching no-store, private, max-age=0, and errors; are we just trying to cache everything? Are there any things that we want to exclude?

Yeah, we are caching pretty much everything we can with this change. I can't really think of anything we want to explicitly exclude. This change just makes it so that if the device goes offline or the upstream server becomes unavailable we'll display cached content if it exists. 

I think this is an ok default for us, especially since some hosting providers (Vercel) don't allow you to set cache-control headers in certain scenarios. 

We could consider exposing a way for apps to customize their caching behaviour in the future, but we'd need to do figure out a way to do that without restarting nginx I think. 

> I'm wondering what the retry interval looks like for 5xx and 429 errors. If there's a single intermittent error how long does it take to recover and retry?

Right now we're relying on the fact that most presentations are in a playlist with more than one presentation so it will eventually be retried once the playlist loops around. 

If there's a single presentation in the playlist then it is a bit tricky because we don't know the presentation had an error. In this case though most apps already need to handle polling for updates but there may be some edge cases where this could cause problems. 

I'll have to think about this a bit more but ideally we could know if the upstream server error'd even if we receive a 200 response from NGINX, allowing us to try reloading again later. Good question!

**Comment by brietaylor** [comment]
> Right now we're relying on the fact that most presentations are in a playlist with more than one presentation so it will eventually be retried once the playlist loops around.

If we cache errors, doesn't this mean that we'll just show a cached error every time we go around the loop, even if the server has recovered in the background, until the cache expires and we can try again?

**Comment by jschr** [comment]
> If we cache errors, doesn't this mean that we'll just show a cached error every time we go around the loop, even if the server has recovered in the background, until the cache expires and we can try again?

If a 429 happens upstream nginx will serve the most recent 200 response (tested on my device with a pre-release version of these changes).  So it will show a stale 200 response until the server recovers.

**Comment by brietaylor** [comment]
Ah, OK, I misunderstood the changes then.  Yes, that sounds good!

**Comment by jschr** [comment]
> Ah, OK, I misunderstood the changes then. Yes, that sounds good!

My PR description was misleading, updated!

**Pull request merged**